lib/temporal/workflow/dispatcher.rb

```diff
@@ -22,11 +37,24 @@ def dispatch(target, event_name, args = nil)
 
       attr_reader :handlers
 
-      def handlers_for(target, event_name)
+      def handlers_for(target, event_name, handler_name)
+        named_handler = sanitize_handler_name(event_name, handler_name)
+
         handlers[target]
           .concat(handlers[TARGET_WILDCARD])
-          .select { |(name, _)| name == event_name || name == WILDCARD }
-          .map(&:last)
+          .select { |event_struct| match?(event_struct, event_name, named_handler) }
+          .map(&:handler)
+      end
+
+      def match?(event_struct, event_name, named_handler)
+        event_struct.event_name == event_name ||
+          event_struct.event_name == named_handler ||
+          event_struct.event_name == WILDCARD
+      end
+
+      def sanitize_handler_name(event_name, handler_name)
```

Comment:
I wonder if we should let the called to this instead of handling it in the Dispatcher ^. Basically I'm thinking that we can leave the Dispatcher's API unchanged and just allow the caller to construct the event name the way they want.

Another approach might be to actually use the target to represent a signal itself, then what you can do is:

```ruby
signal = Signal.new(signal_name)
dispatcher.register_handler(signal, 'signaled') { |input| ... }
```

and for a catch-all:

```ruby
dispatcher.register_handler(Dispatcher::TARGET_WILDCARD, 'signaled') { |signal, input| ... }
```

I'm a bit hesitant to extend the Dispatcher's interface because I feel like we can achieve what we need with the current version and can avoid overcomplicating it

Reply:
As you see, I tried to move the naming logic into Dispatcher. I agree that it changes the Dispatcher interface but now the complexity will move elsewhere.

I like the second suggestion about making a `Signal` object. I'll try that first since it preserves some encapsulation without leaking it to multiple classes.